File: solutions/csharp/wordy/1/Wordy.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using Sprache;

public static class Wordy
{
    private static class Grammer
    {
        private static readonly Parser<string> WhatIs = Parse.String("What is").Text().Token();
        private static readonly Parser<char> QuestionMark = Parse.Char('?').Token();
        private static readonly Parser<Expression> Constant =
            from op in Parse.Optional(Parse.Char('-').Token())
            from n in Parse.Number.Token()
            select Expression.Constant(int.Parse(n) * (op.IsDefined ? -1 : 1));
        private static readonly Parser<ExpressionType> Operator =
            Parse.String("plus").Return(ExpressionType.Add)
                .Or(Parse.String("minus").Return(ExpressionType.Subtract))
                .Or(Parse.String("multiplied by").Return(ExpressionType.Multiply))
                .Or(Parse.String("divided by").Return(ExpressionType.Divide));
        private static readonly Parser<Expression> Operation =
            Parse.ChainOperator(Operator, Constant, Expression.MakeBinary);

        public static readonly Parser<Expression> Question =
            from whatIs in WhatIs
            from expression in Operation.Or(Constant)
            from questionMark in QuestionMark.End()
            select expression;
    }

    public static int Answer(string question)
    {
        try
        {
            var operation = Grammer.Question.Parse(question);
            var func = Expression.Lambda<Func<int>>(operation).Compile();
            return func();
        }
        catch (Exception e)
        {
            throw new ArgumentException("Invalid input", e);
        }
    }
}

File: solutions/go/wordy/1/wordy.go
package wordy

import (
	"github.com/alecthomas/participle"
)

type operator int

const (
	plus operator = iota
	minus
	times
	divide
)

var operatorMap = map[string]operator{"plus": plus, "minus": minus, "multiplied": times, "divided": divide}

func (o *operator) Capture(s []string) error {
	*o = operatorMap[s[0]]
	return nil
}

type term struct {
	Value *int `parser:"@('-'? Int)"`
}

type opTerm struct {
	Operator operator `@("plus" | "minus" | "multiplied" "by" | "divided" "by")`
	Term     *term    `@@`
}

type experssion struct {
	Term   *term     `@@`
	OpTerm []*opTerm `@@*`
}

type question struct {
	WhatIs       string      `"What" "is"`
	Expression   *experssion `@@`
	QuestionMark string      `"?"`
}

// Evaluation
func (o operator) eval(l, r int) int {
	switch o {
	case plus:
		return l + r
	case minus:
		return l - r
	case times:
		return l * r
	case divide:
		return l / r
	}
	panic("unsupported operator")
}

func (t *term) eval() int {
	return *t.Value
}

func (ot *opTerm) eval(l int) int {
	r := ot.Term.eval()
	return ot.Operator.eval(l, r)
}

func (e *experssion) eval() int {
	v := e.Term.eval()
	for _, ot := range e.OpTerm {
		v = ot.eval(v)
	}
	return v
}

func (q *question) eval() int {
	return q.Expression.eval()
}

// Answer evaluates question and returns result if successful
func Answer(q string) (int, bool) {
	parser, err := participle.Build(&question{})
	if err != nil {
		return 0, false
	}

	question := &question{}
	err = parser.ParseString(q, question)
	if err != nil {
		return 0, false
	}

	a := question.eval()
	return a, true
}
